src/io/unix/sync.rs

```diff
@@ -59,39 +59,50 @@ impl File {
         self.size.load(Ordering::SeqCst)
     }
 
-    pub(crate) async fn write_append_all(&self, buf: Bytes) -> IOResult<()> {
-        let offset = self.size.fetch_add(buf.len() as u64, Ordering::SeqCst);
-        self.write_all_at(offset, buf).await
-    }
-
-    pub(crate) async fn write_append_all_buffers(
+    pub(crate) async fn write_append_writable_data<R: Send + 'static>(
         &self,
-        first_buf: Bytes,
-        second_buf: Bytes,
-    ) -> IOResult<()> {
-        let first_len = first_buf.len() as u64;
-        let second_len = second_buf.len() as u64;
-        let len = first_len + second_len;
+        c: impl WritableDataCreator<R>,
+        len: u64,
+    ) -> Result<R> {
         let mut offset = self.size.fetch_add(len, Ordering::SeqCst);
         let file = self.no_lock_fd.clone();
         if Self::can_run_inplace(len) {
             Self::inplace_sync_call(move || {
-                file.write_all_at(&first_buf, offset)?;
-                offset = offset + first_len;
-                file.write_all_at(&second_buf, offset)?;
-                Ok(())
+                let (res, data) = c(offset)?;
+                match res {
+                    WritableData::Single(bytes) => file.write_all_at(&bytes, offset)?,
+                    WritableData::Double(b1, b2) => {
+                        file.write_all_at(&b1, offset)?;
+                        offset = offset + b1.len() as u64;
+                        file.write_all_at(&b2, offset)?;
+                    }
+                };
+                Ok(data)
             })
         } else {
             Self::background_sync_call(move || {
-                file.write_all_at(&first_buf, offset)?;
-                offset = offset + first_len;
-                file.write_all_at(&second_buf, offset)?;
-                Ok(())
+                let (res, data) = c(offset)?;
+                match res {
+                    WritableData::Single(bytes) => {
+                        file.write_all_at(&bytes, offset)?;
+                    }
+                    WritableData::Double(b1, b2) => {
+                        file.write_all_at(&b1, offset)?;
+                        offset = offset + b1.len() as u64;
+                        file.write_all_at(&b2, offset)?;
+                    }
+                };
+                Ok(data)
             })
             .await
         }
     }
 
+    pub(crate) async fn write_append_all(&self, buf: Bytes) -> IOResult<()> {
+        let offset = self.size.fetch_add(buf.len() as u64, Ordering::SeqCst);
+        self.write_all_at(offset, buf).await
```

Comment:
I think it would  be good to update this function as well: perform `self.size.fetch_add` within the code block passed to `inplace_sync_call` or to `background_sync_call`. This will require to manually inline the code from `write_all_at` to this function, but I think this is not a big problem